node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/HostDeprovisionMaintainer.java

```diff
@@ -0,0 +1,65 @@
+// Copyright 2019 Oath Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.provision.maintenance;
+
+import com.yahoo.config.provision.NodeType;
+import com.yahoo.transaction.Mutex;
+import com.yahoo.vespa.hosted.provision.Node;
+import com.yahoo.vespa.hosted.provision.NodeList;
+import com.yahoo.vespa.hosted.provision.NodeRepository;
+import com.yahoo.vespa.hosted.provision.provisioning.HostProvisioner;
+
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * @author freva
+ */
+public class HostDeprovisionMaintainer extends Maintainer {
+
+    private static final Logger log = Logger.getLogger(HostDeprovisionMaintainer.class.getName());
+
+    private final HostProvisioner hostProvisioner;
+
+    public HostDeprovisionMaintainer(
+            NodeRepository nodeRepository, Duration interval, JobControl jobControl, HostProvisioner hostProvisioner) {
+        super(nodeRepository, interval, jobControl);
+        this.hostProvisioner = hostProvisioner;
+    }
+
+    @Override
+    protected void maintain() {
+        try (Mutex lock = nodeRepository().lockUnallocated()) {
+            NodeList nodes = nodeRepository().list();
+
+            for (Node node : actionableNodes(nodes)) {
+                try {
+                    hostProvisioner.deprovision(node);
+                    nodeRepository().removeRecursively(node.hostname());
+                } catch (RuntimeException e) {
+                    log.log(Level.WARNING, "Failed to deprovision " + node.hostname() + ", will retry in " + interval(), e);
+                }
+            }
+        }
+    }
+
+    /** @return Nodes of type host, in any state, that have no children with allocation */
```

Comment:
I think it would good to verify that nodes are managed by the provision service here, so that the guarantee that we do not deprovision unexpected nodes can be here and not in the `HostProvisioner` implementation. How about a flag on the node object? E.g `provisionedDynamically`.

Reply:
Is this a safety precaution or a technical reason?
Because on the technical side, there is no difference between a host provisioned manually and provisioned with provision service. The provision service will be able to deprovision any host.
Therefore, anything that this method returns will be deprovisioned and there are no additional checks in the provision service. In that regard, there shouldn't be anything unexpected.

This flag, perhaps named differently, could still be added to ensure a pool of hosts that are not touched by `HostProvisioner`, but that will greatly increase complexity. F.ex. if dynamic allocation is not allowed, that pool will be quickly exhausted anyway. If it is, then how to decide what is dynamically allocated and what is dynamically provisioned? How do we keep in sync who the owner is of the different hardware?

Reply:
> there is no difference between a host provisioned manually and provisioned with provision service. The provision service will be able to deprovision any host.

Right. All hosts will be either dynamically provisioned or not, in a given zone. Never mind 🙂